File: src/robot/Robot.ts
import type {
  Position,
  RobotAction,
  RobotCommand,
  RobotConfig,
  RobotState,
} from '../types/index.js';
import { RobotActionType } from '../types/index.js';
import { Vector2D } from '../utils/Vector2D.js';

/**
 * Create initial robot state from configuration
 */
export function createRobotState(
  config: RobotConfig,
  alliance: 'red' | 'blue',
  startPosition: Position,
  startHeading: number = 0
): RobotState {
  return {
    id: config.id,
    config,
    alliance,
    position: { ...startPosition },
    heading: startHeading,
    velocity: 0,
    heldBalls: [],
    currentAction: {
      type: RobotActionType.IDLE,
      progress: 0,
      startedAt: 0,
    },
    disabled: false,
    hasAutoClimbed: false,
    hasClimbed: false,
    currentClimbLevel: null,
    currentPath: [],
    pathIndex: 0,
  };
}

/**
 * Robot entity class for managing robot state and operations
 */
export class Robot {
  state: RobotState;

  constructor(
    config: RobotConfig,
    alliance: 'red' | 'blue',
    startPosition: Position,
    startHeading: number = 0
  ) {
    this.state = createRobotState(config, alliance, startPosition, startHeading);
  }

  get id(): string {
    return this.state.id;
  }

  get config(): RobotConfig {
    return this.state.config;
  }

  get alliance(): 'red' | 'blue' {
    return this.state.alliance;
  }

  get position(): Position {
    return this.state.position;
  }

  get heading(): number {
    return this.state.heading;
  }

  get velocity(): number {
    return this.state.velocity;
  }

  get heldBalls(): string[] {
    return this.state.heldBalls;
  }

  get currentAction(): RobotAction {
    return this.state.currentAction;
  }

  get isDisabled(): boolean {
    return this.state.disabled;
  }

  get hasClimbed(): boolean {
    return this.state.hasClimbed;
  }

  get hasAutoClimbed(): boolean {
    return this.state.hasAutoClimbed;
  }

  get currentClimbLevel(): number | null {
    return this.state.currentClimbLevel;
  }

  /**
   * Check if robot is currently idle
   */
  isIdle(): boolean {
    return this.state.currentAction.type === RobotActionType.IDLE;
  }

  /**
   * Check if robot is currently moving
   */
  isMoving(): boolean {
    return this.state.currentAction.type === RobotActionType.MOVING;
  }

  /**
   * Check if robot can pick up more balls
   */
  canPickUpBall(): boolean {
    return this.state.heldBalls.length < this.config.ballCapacity;
  }

  /**
   * Check if robot has balls to shoot
   */
  hasBalls(): boolean {
    return this.state.heldBalls.length > 0;
  }

  /**
   * Get the number of balls held
   */
  ballCount(): number {
    return this.state.heldBalls.length;
  }

  /**
   * Pick up a ball
   */
  pickUpBall(ballId: string): boolean {
    if (!this.canPickUpBall()) return false;
    this.state.heldBalls.push(ballId);
    return true;
  }

  /**
   * Remove and return a ball for shooting
   */
  shootBall(): string | null {
    if (this.state.heldBalls.length === 0) return null;
    return this.state.heldBalls.shift() ?? null;
  }

  /**
   * Set the current action
   */
  setAction(action: RobotAction): void {
    this.state.currentAction = action;
  }

  /**
   * Start a new action from a command
   */
  startAction(command: RobotCommand, tick: number): void {
    this.state.currentAction = {
      type: command.type,
      targetPosition: command.targetPosition,
      targetBallId: command.targetBallId,
      targetScoringZoneId: command.targetScoringZoneId,
      targetRobotId: command.targetRobotId,
      targetClimbLevel: command.targetClimbLevel,
      isAutoClimb: command.isAutoClimb,
      progress: 0,
      startedAt: tick,
    };
  }

  /**
   * Update action progress
   */
  updateProgress(progress: number): void {
    this.state.currentAction.progress = Math.min(1, progress);
  }

  /**
   * Complete current action and go idle
   */
  completeAction(): void {
    this.state.currentAction = {
      type: RobotActionType.IDLE,
      progress: 0,
      startedAt: 0,
    };
  }

  /**
   * Set position
   */
  setPosition(pos: Position): void {
    this.state.position = { ...pos };
  }

  /**
   * Set heading
   */
  setHeading(heading: number): void {
    this.state.heading = ((heading % 360) + 360) % 360;
  }

  /**
   * Set velocity
   */
  setVelocity(velocity: number): void {
    this.state.velocity = Math.max(0, velocity);
  }

  /**
   * Set path for navigation
   */
  setPath(path: Position[]): void {
    this.state.currentPath = path;
    this.state.pathIndex = 0;
  }

  /**
   * Advance to next path waypoint
   */
  advancePath(): Position | null {
    if (this.state.pathIndex >= this.state.currentPath.length - 1) {
      return null;
    }
    this.state.pathIndex++;
    return this.state.currentPath[this.state.pathIndex];
  }

  /**
   * Get current path target
   */
  getCurrentPathTarget(): Position | null {
    if (this.state.pathIndex >= this.state.currentPath.length) {
      return null;
    }
    return this.state.currentPath[this.state.pathIndex];
  }

  /**
   * Check if path is complete
   */
  isPathComplete(): boolean {
    if (this.state.currentPath.length === 0) return true;
    return this.state.pathIndex >= this.state.currentPath.length - 1;
  }

  /**
   * Disable the robot
   */
  disable(): void {
    this.state.disabled = true;
    this.state.velocity = 0;
    this.completeAction();
  }

  /**
   * Enable the robot
   */
  enable(): void {
    this.state.disabled = false;
  }

  /**
   * Mark robot as climbed during auto
   */
  autoClimb(): void {
    this.state.hasAutoClimbed = true;
    this.state.velocity = 0;
  }

  /**
   * Mark robot as climbed in endgame with a specific level
   */
  endgameClimb(level: number): void {
    this.state.hasClimbed = true;
    this.state.currentClimbLevel = level;
    this.state.velocity = 0;
  }

  /**
   * @deprecated Use autoClimb() or endgameClimb(level) instead
   * Mark robot as climbed (legacy method)
   */
  climb(): void {
    this.endgameClimb(1);
  }

  /**
   * Get distance to a position
   */
  distanceTo(pos: Position): number {
    return Vector2D.fromPosition(this.position).distanceTo(pos);
  }

  /**
   * Get angle to a position
   */
  angleTo(pos: Position): number {
    return Vector2D.fromPosition(this.position).angleTo(pos);
  }

  /**
   * Get bounding box corners
   */
  getBoundingBox(): { minX: number; maxX: number; minY: number; maxY: number } {
    const halfWidth = this.config.width / 2;
    const halfLength = this.config.length / 2;
    return {
      minX: this.position.x - halfWidth,
      maxX: this.position.x + halfWidth,
      minY: this.position.y - halfLength,
      maxY: this.position.y + halfLength,
    };
  }

  /**
   * Clone the robot state
   */
  cloneState(): RobotState {
    return {
      ...this.state,
      position: { ...this.state.position },
      heldBalls: [...this.state.heldBalls],
      currentAction: { ...this.state.currentAction },
      currentPath: this.state.currentPath.map((p) => ({ ...p })),
    };
  }
}

File: python_server/frc_sim_server/entities/robot.py
"""Robot entity class."""

from __future__ import annotations
import uuid
from typing import Optional
from dataclasses import dataclass, field

from ..types.enums import RobotActionType
from ..types.schemas import Position, RobotConfig, RobotAction, RobotState


def create_idle_action(tick: int = 0) -> RobotAction:
    """Create an idle action."""
    return RobotAction(
        type=RobotActionType.IDLE.value,
        progress=0.0,
        startedAt=tick,
    )


@dataclass
class Robot:
    """Robot entity with state and behavior."""

    config: RobotConfig
    alliance: str
    position: Position
    heading: float
    velocity: float = 0.0
    held_balls: list[str] = field(default_factory=list)
    current_action: RobotAction = field(default_factory=lambda: create_idle_action())
    secondary_action: Optional[RobotAction] = None
    is_disabled: bool = False
    has_auto_climbed: bool = False
    has_climbed: bool = False
    current_climb_level: Optional[int] = None
    current_path: list[Position] = field(default_factory=list)
    path_index: int = 0

    @property
    def id(self) -> str:
        """Get robot ID."""
        return self.config.id

    def clone_state(self) -> RobotState:
        """Clone the robot state for serialization."""
        return RobotState(
            id=self.config.id,
            config=self.config,
            alliance=self.alliance,
            position=self.position,
            heading=self.heading,
            velocity=self.velocity,
            heldBalls=list(self.held_balls),
            currentAction=self.current_action,
            secondaryAction=self.secondary_action,
            disabled=self.is_disabled,
            hasAutoClimbed=self.has_auto_climbed,
            hasClimbed=self.has_climbed,
            currentClimbLevel=self.current_climb_level,
            currentPath=list(self.current_path),
            pathIndex=self.path_index,
        )

    def is_idle(self) -> bool:
        """Check if robot is idle."""
        return self.current_action.type == RobotActionType.IDLE.value

    def is_moving(self) -> bool:
        """Check if robot is moving."""
        return self.current_action.type == RobotActionType.MOVING.value

    def is_shooting(self) -> bool:
        """Check if robot is shooting."""
        return self.current_action.type == RobotActionType.SHOOTING.value

    def has_balls(self) -> bool:
        """Check if robot has any balls."""
        return len(self.held_balls) > 0

    def can_pick_up_ball(self) -> bool:
        """Check if robot can pick up another ball."""
        return len(self.held_balls) < self.config.ballCapacity

    def has_secondary_action(self) -> bool:
        """Check if robot has a secondary action in progress."""
        return self.secondary_action is not None

    def distance_to(self, pos: Position) -> float:
        """Calculate distance to a position."""
        dx = pos.x - self.position.x
        dy = pos.y - self.position.y
        return (dx * dx + dy * dy) ** 0.5

    def set_position(self, pos: Position) -> None:
        """Set robot position."""
        self.position = pos

    def set_heading(self, heading: float) -> None:
        """Set robot heading."""
        self.heading = heading

    def set_velocity(self, velocity: float) -> None:
        """Set robot velocity."""
        self.velocity = velocity

    def set_path(self, path: list[Position]) -> None:
        """Set the path for the robot to follow."""
        self.current_path = path
        self.path_index = 0

    def get_current_path_target(self) -> Optional[Position]:
        """Get the current target position on the path."""
        if self.path_index < len(self.current_path):
            return self.current_path[self.path_index]
        return None

    def advance_path(self) -> Optional[Position]:
        """Advance to the next path waypoint."""
        self.path_index += 1
        return self.get_current_path_target()

    def start_action(self, action: RobotAction, tick: int) -> None:
        """Start a new action."""
        self.current_action = RobotAction(
            type=action.type,
            progress=0.0,
            startedAt=tick,
            targetPosition=action.targetPosition,
            targetBallId=action.targetBallId,
            targetBallIds=action.targetBallIds,
            targetScoringZoneId=action.targetScoringZoneId,
            targetRobotId=action.targetRobotId,
            targetClimbLevel=action.targetClimbLevel,
            isAutoClimb=action.isAutoClimb,
        )

    def update_progress(self, progress: float) -> None:
        """Update action progress."""
        self.current_action = RobotAction(
            type=self.current_action.type,
            progress=progress,
            startedAt=self.current_action.startedAt,
            targetPosition=self.current_action.targetPosition,
            targetBallId=self.current_action.targetBallId,
            targetBallIds=self.current_action.targetBallIds,
            targetScoringZoneId=self.current_action.targetScoringZoneId,
            targetRobotId=self.current_action.targetRobotId,
            targetClimbLevel=self.current_action.targetClimbLevel,
            isAutoClimb=self.current_action.isAutoClimb,
        )

    def complete_action(self) -> None:
        """Complete the current action."""
        self.current_action = create_idle_action()
        self.current_path = []
        self.path_index = 0

    def pick_up_ball(self, ball_id: str) -> None:
        """Pick up a ball."""
        if ball_id not in self.held_balls:
            self.held_balls.append(ball_id)

    def shoot_ball(self) -> Optional[str]:
        """Shoot a ball and return its ID."""
        if self.held_balls:
            return self.held_balls.pop(0)
        return None

    def start_secondary_pickup(self, ball_id: str, tick: int) -> None:
        """Start a secondary pickup action."""
        self.secondary_action = RobotAction(
            type=RobotActionType.PICKING_UP.value,
            progress=0.0,
            startedAt=tick,
            targetBallId=ball_id,
        )

    def update_secondary_progress(self, progress: float) -> None:
        """Update secondary action progress."""
        if self.secondary_action:
            self.secondary_action = RobotAction(
                type=self.secondary_action.type,
                progress=progress,
                startedAt=self.secondary_action.startedAt,
                targetBallId=self.secondary_action.targetBallId,
            )

    def complete_secondary_action(self) -> None:
        """Complete the secondary action."""
        self.secondary_action = None

    def auto_climb(self) -> None:
        """Mark robot as having auto-climbed."""
        self.has_auto_climbed = True

    def endgame_climb(self, level: int) -> None:
        """Mark robot as having climbed in endgame."""
        self.has_climbed = True
        self.current_climb_level = level
